### PR TITLE
feat(`DeleteSourceDialog`): when lots of sources are selected for deletion, block the continue button for 5 seconds

### LINKED ISSUE #2297: Add a timer to delay activation of the delete dialog's "YES,..." button when multiple sources are selected
(either all sources or more sources than can be comfortably displayed in the dialog)

### PR DESCRIPTION
**Pull request by cfm**
## Status

Ready for review

## Description

Closes #2297 by blocking the `DeleteSourceDialog`'s "continue" button for 5 seconds, with a countdown timer.

## Test Plan

This can be tested easily with `make dev` with:

```patch
--- a/client/securedrop_client/gui/source/delete/dialog.py
+++ b/client/securedrop_client/gui/source/delete/dialog.py
@@ -27,7 +27,7 @@ from securedrop_client.gui.base import ModalDialog
 
 # Maximum number of source names to display in delete dialog before
 # (a) truncation of the list and (b) delayed confirmation:
-LOTS_OF_SOURCES = 30
+LOTS_OF_SOURCES = 2
 
 # Timers for delayed confirmation:
 SEC = 1000  # ms
```

Then:

1. [ ] With 1 or 2 sources selected, the `Yes, delete N source accounts button` is immediately enabled and works as expected.
2. [ ] With all 3 sources selected, the `Yes, delete N source accounts button`:
    1. [ ] is initially disabled;
    2. [ ] counts down from 5 seconds; and then
    3. [ ] becomes enabled and works as expected.

### CODE REVIEW

**Comment by cfm** [comment]
NB. Generalizing this UX for `ModalDialog(dangerous=VERY)` is left for future work.

**Comment by zenmonkeykstop** [comment]
This change looks great, happy to review for reals once tests in place.

**Comment by cfm** [comment]
Thanks, @zenmonkeykstop.  Functional test added in 224e91a747fc4acea9e7cf72cd96666452b60ded.

**Review by zenmonkeykstop — APPROVED** [review]
Test plan checks out, liking the countdown.

**Comment by veteran-2022** [comment]
@CodiumAI-Agent /review

**Comment by veteran-2022** [comment]
@CodiumAI-Agent

**Comment by veteran-2022** [comment]
@Veteran-ChengQin /test-plan

**Comment by veteran-2022** [comment]
@Veteran-ChengQin /test-plan

**Comment by veteran-2022** [comment]
@Veteran-ChengQin /test-plan

**Comment by Veteran-ChengQin** [comment]
# Test Plan

# Test Plan for PR: feat(`DeleteSourceDialog`): when lots of sources are selected for deletion, block the continue button for 5 seconds

## 1. Purpose
This test plan verifies the addition of a feature to the `DeleteSourceDialog` that disables the "continue" button for 5 seconds when a significant number of sources are selected for deletion. The functionality includes a countdown timer displayed alongside the button. The plan also ensures the new feature integrates with the existing application UI seamlessly and does not introduce regressions or inconsistencies.

## 2. Scope
### In Scope:
- `DeleteSourceDialog` class behavior around the 5-second delay feature.
- Countdown timer functionality.
- Integration of the delayed confirmation feature with existing deletion workflows.
- Check the `continue_button` state transitions (disabled/enabled) based on the countdown timer.
- Ensure the UI updates correctly while sources are selected for deletion.
- Regression testing for the functionality that deletes sources.

### Out of Scope:
- Other unrelated features of `DeleteSourceDialog`.
- Localization of text in languages not currently supported in the project.

## 3. Test Environment
- OS: Linux/Windows/macOS (as applicable for SecureDrop clients)
- Frameworks: PyQt5
- Tools: qtbot (for GUI interaction), flaky (for ensuring stable test runs)
- SecureDrop client application with mock data from test configurations.

## 4. Test Cases

### Test Case 1: Verify `DeleteSourceDialog` Blocks Continue Button for 5 Seconds
- **Test Objective**: Verify that the `continue_button` remains disabled for the specified time when more than `LOTS_OF_SOURCES` are selected for deletion.
- **Preconditions**:
    - Set `LOTS_OF_SOURCES` to a specific threshold.
    - Sources are already synced and available in the source list UI.
- **Test Steps**:
    1. Open the SecureDrop client.
    2. Select the first source using a regular click.
    3. Select subsequent sources using `Ctrl` + click, ensuring the total count exceeds the `LOTS_OF_SOURCES` threshold.
    4. Trigger the deletion action via the batch actions toolbar.
    5. Observe the `DeleteSourceDialog` appears with the `continue_button` disabled and a countdown timer displayed.
    6. Wait for 5 seconds until the countdown completes.
    7. Check if the `continue_button` becomes enabled once the timer expires.
- **Expected Results**:
    - The `continue_button` is disabled for 5 seconds with a visible countdown timer.
    - The button re-enables after the specified time and displays its original text.
    - The sources can only be deleted after the delay expires.

---

### Test Case 2: Verify Countdown Timer Updates Correctly
- **Test Objective**: Ensure the countdown timer updates correctly, reflecting the time left for the `continue_button` to re-enable.
- **Preconditions**:
    - Multiple sources are selected for deletion.
    - `LOTS_OF_SOURCES` count is exceeded.
- **Test Steps**:
    1. Trigger the deletion action for multiple sources exceeding the `LOTS_OF_SOURCES` threshold.
    2. Observe the countdown timer adjacent to the `continue_button`.
    3. Verify the timer decrements every second until reaching zero.
- **Expected Results**:
    - The timer displays accurate remaining time.
    - The countdown completes smoothly and re-enables the `continue_button`.

---

### Test Case 3: Verify UI Behavior with Mixed Source Selection
- **Test Objective**: Ensure the `DeleteSourceDialog` behaves correctly when a mix of source types is selected.
- **Preconditions**:
    - The source list contains various sources synced from the database.
- **Test Steps**:
    1. Select a mix of flagged and unflagged sources.
    2. Trigger the deletion action for sources exceeding the `LOTS_OF_SOURCES` threshold.
    3. Observe the dialog and the behavior of the `continue_button` and confirmation text.
- **Expected Results**:
    - The dialog correctly lists selected sources in the confirmation text.
    - The `continue_button` remains disabled for 5 seconds and re-enables after the timer expires.

---

### Test Case 4: Verify Functional Tests
- **Test Objective**: Execute functional tests added in `test_delete_sources.py` and `test_confirm_before_deleting_lots_of_sources.yml` for verifying the new functionality.
- **Preconditions**:
    - Added functional tests are set up with sufficient mock data.
- **Test Steps**:
    1. Run the functional tests from `test_delete_sources.py`.
    2. Review output for `test_confirm_before_deleting_lots_of_sources`.
    3. Ensure tests for both timer delay behavior and source handling pass.
- **Expected Results**:
    - Functional tests pass with no errors or failures.

---

### Test Case 5: Verify Regression for Existing Deletion Workflow
- **Test Objective**: Ensure the new countdown timer does not break existing deletion workflows for fewer sources.
- **Preconditions**:
    - Select fewer than `LOTS_OF_SOURCES`.
- **Test Steps**:
    1. Select a small number of sources for deletion.
    2. Trigger the deletion action.
    3. Observe the `DeleteSourceDialog` behaves without blocking the `continue_button`.
- **Expected Results**:
    - The dialog allows immediate deletion of sources without a timer delay.
---

## 5. Special Considerations
- **Edge Cases**:
    - Selecting exactly `LOTS_OF_SOURCES` and verifying dialog behavior.
    - Exceeding `LOTS_OF_SOURCES` by a significant margin, e.g., hundreds of sources.
    - Simulated errors in timer updates during countdown.
- **Risks**:
    - UI performance during large batch selections.
    - Functional regressions in source list updates after deletion.
- **Dependencies**:
    - Test cases rely on accurate mock data and proper mocking of database entries.

*自动生成的测试计划*

**Comment by veteran-2022** [comment]
@Veteran-ChengQin /test-plan

**Comment by veteran-2022** [comment]
@Veteran-ChengQin /test-plan

**Comment by veteran-2022** [comment]
@Veteran-ChengQin /test-plan

**Comment by Veteran-ChengQin** [comment]
# Test Plan

# Test Plan for PR: feat(`DeleteSourceDialog`): when lots of sources are selected for deletion, block the continue button for 5 seconds

## 1. Purpose

This test plan ensures that the new functionality implemented in the `DeleteSourceDialog` works as intended. Specifically, when many sources are selected for deletion, the "continue" button is disabled for 5 seconds with a visible countdown timer. The test plan includes functional testing of the new feature, edge case scenarios, and regression testing for related functionalities.

## 2. Scope

### In Scope:
- Validation of the countdown timer functionality for the "continue" button when many sources are selected for deletion.
- GUI updates, including the disabling/enabling of the "continue" button and updating its text to reflect the time delay.
- Event handling (timer mechanism) to dynamically update the button state.
- Integration with the larger `DeleteSourceDialog` behavior.

### Out of Scope:
- Non-GUI functionalities unrelated to this new timer behavior.
- Low-level testing of other unrelated modules or dialogs in the application.

## 3. Test Environment

- **OS:** Compatible with the development and production environment platforms.
- **Frameworks/Dependencies:**
  - PyQt5 (for GUI components)
  - Functional testing tools (e.g., pytest, qtbot)
  - Flaky library for robust testing of intermittent functionality
- **Configuration:** Mock or actual database with source records.
- **Preconditions:**
  - The environment should have a list of sources to delete.
  - The application must be functional and fully integrated with all required dependencies.

## 4. Test Cases

### Test Case 1: Standard Countdown Timer
- **Objective:** Verify that the "continue" button in the `DeleteSourceDialog` is disabled for 5 seconds when more than 30 sources are selected.
- **Preconditions:** The `LOTS_OF_SOURCES` threshold is set to 30.
- **Steps:**
  1. Open the `DeleteSourceDialog` with more than 30 sources selected.
  2. Observe the initial state of the "continue" button and its displayed text.
  3. Wait for 5 seconds.
  4. Observe the final state of the "continue" button and its displayed text.
- **Expected Results:**
  - Initially, the "continue" button is disabled, and its text includes a countdown (e.g., "wait 5 sec").
  - After 5 seconds, the button is re-enabled, and the original text appears.
- **Priority:** High

### Test Case 2: Boundary Values for Source Count
- **Objective:** Verify that the timer behavior only triggers when the number of sources exceeds the `LOTS_OF_SOURCES` threshold.
- **Preconditions:** The `LOTS_OF_SOURCES` threshold is set to 30.
- **Steps:**
  1. Open the `DeleteSourceDialog` with exactly 30 sources selected.
  2. Confirm that the button is not delayed and is immediately enabled.
  3. Open the `DeleteSourceDialog` with 31 sources selected.
  4. Confirm that the button delay behavior is triggered.
- **Expected Results:**
  - With 30 sources, the "continue" button is immediately enabled.
  - With 31 sources, the "continue" button is disabled and follows the timer behavior.
- **Priority:** High

### Test Case 3: Edge Case - Dialog Cancelation During Countdown
- **Objective:** Verify that canceling the dialog before the countdown finishes stops the timer and restores the initial state.
- **Preconditions:** The `DeleteSourceDialog` is opened with more than 30 sources.
- **Steps:**
  1. Open the `DeleteSourceDialog`.
  2. Cancel the dialog while the timer is active.
  3. Re-open the `DeleteSourceDialog` with the same number of sources.
- **Expected Results:**
  - The timer is reset and does not continue from the previous state.
  - The dialog behaves as expected upon reopening.
- **Priority:** Medium

### Test Case 4: Edge Case - Rapid Dialog Reopen
- **Objective:** Test if reopening the dialog immediately after closing it maintains the correct timer behavior.
- **Preconditions:** The `DeleteSourceDialog` is opened with a significant number of sources.
- **Steps:**
  1. Open and immediately close the `DeleteSourceDialog`.
  2. Re-open it with the same or different number of sources.
- **Expected Results:**
  - The timer restarts each time the dialog is reopened.
  - No residual issues (e.g., frozen GUI or incorrect button state) occur.
- **Priority:** Medium

### Test Case 5: Regression Test - Standard "DeleteSourceDialog" Behavior
- **Objective:** Ensure that the `DeleteSourceDialog` behaves as intended when less than the threshold number of sources are selected.
- **Preconditions:** The `DeleteSourceDialog` is opened with fewer than 30 sources.
- **Steps:**
  1. Open the dialog with fewer sources.
  2. Verify that the "continue" button is immediately enabled.
  3. Complete the deletion and validate the state of the remaining sources.
- **Expected Results:**
  - The dialog works correctly without delays or GUI issues.
  - The "delete" action completes as expected.
- **Priority:** High

### Test Case 6: Timer Accuracy
- **Objective:** Verify that the countdown timer text accurately reflects the remaining time.
- **Preconditions:** The `DeleteSourceDialog` is opened with more than 30 sources.
- **Steps:**
  1. Open the dialog and observe the timer countdown text.
  2. Confirm that the remaining time displayed decreases each second until the timer expires.
- **Expected Results:**
  - The timer countdown text updates every second without skips or inaccuracies.
- **Priority:** Medium

## 5. Special Considerations
- Ensure robust handling of edge cases like quick reopening or closing of the dialog.
- Validate that the GUI updates (button state, text) adhere to accessibility guidelines.
- Perform regression testing on other parts of the application affected by this dialog.
- Consider performance tests if opening the dialog with many sources causes delays.

*自动生成的测试计划*

**Comment by legoktm** [comment]
@veteran-2022 this PR was merged 4 months ago, there's no need to be testing AI tools against it. Can you explain what you're trying to do?

**Comment by Veteran-ChengQin** [comment]
> [CodiumAI-Agent](/CodiumAI-Agent)

Hello, I am testing a test plan agent we developed. Please forgive me if it causes you any trouble.

**Comment by Veteran-ChengQin** [comment]
> @veteran-2022 this PR was merged 4 months ago, there's no need to be testing AI tools against it. Can you explain what you're trying to do?

Are you interested in helping us test it? We will provide free access.

**Comment by legoktm** [comment]
No, please don't test your AI/developer tools in our repositories.
